olm/operator-controller/v1alpha1/clusterextension_types.go

```diff
@@ -0,0 +1,518 @@
+type SourceConfig struct {
+	// sourceType is a required reference to the type of install source.
+	//
+	// Allowed values are ["Catalog"]
+	//
+	// When this field is set to "Catalog", information for determining the appropriate
+	// bundle of content to install will be fetched from ClusterCatalog resources existing
+	// on the cluster. When using the Catalog sourceType, the catalog field must also be set.
+	//
+	// +unionDiscriminator
+	// +kubebuilder:validation:Enum:="Catalog"
+	SourceType string `json:"sourceType"`
+
+	// catalog is used to configure how information is sourced from a catalog. This field must be defined when sourceType is set to "Catalog",
+	// and must be the only field defined for this sourceType.
+	//
+	// +optional.
+	Catalog *CatalogSource `json:"catalog,omitempty"`
+}
+
+// ClusterExtensionInstallConfig is a union which selects the clusterExtension installation config.
+// ClusterExtensionInstallConfig requires the namespace and serviceAccount which should be used for the installation of packages.
+// +union
+type ClusterExtensionInstallConfig struct {
+	// namespace is a reference to the Namespace in which the bundle of
+	// content for the package referenced in the packageName field will be applied.
+	// The bundle may contain cluster-scoped resources or resources that are
+	// applied to other Namespaces. This Namespace is expected to exist.
+	//
+	// namespace is required, immutable, and follows the DNS label standard
+	// as defined in [RFC 1123]. This means that valid values:
+	//   - Contain no more than 63 characters
+	//   - Contain only lowercase alphanumeric characters or '-'
+	//   - Start with an alphanumeric character
+	//   - End with an alphanumeric character
+	//
+	// Some examples of valid values are:
+	//   - some-namespace
+	//   - 123-namespace
+	//   - 1-namespace-2
+	//   - somenamespace
+	//
+	// Some examples of invalid values are:
+	//   - -some-namespace
+	//   - some-namespace-
+	//   - thisisareallylongnamespacenamethatisgreaterthanthemaximumlength
+	//   - some.namespace
+	//
+	// [RFC 1123]: https://tools.ietf.org/html/rfc1123
+	//
+	//+kubebuilder:validation:Pattern:=^[a-z0-9]([-a-z0-9]*[a-z0-9])?$
+	//+kubebuilder:validation:MaxLength:=63
+	//+kubebuilder:validation:XValidation:rule="self == oldSelf",message="namespace is immutable"
+	Namespace string `json:"namespace"`
+
+	// serviceAccount is a required reference to a ServiceAccount that exists
+	// in the installNamespace. The provided ServiceAccount is used to install and
+	// manage the content for the package specified in the packageName field.
+	//
+	// In order to successfully install and manage the content for the package,
+	// the ServiceAccount provided via this field should be configured with the
+	// appropriate permissions to perform the necessary operations on all the
+	// resources that are included in the bundle of content being applied.
```

Comment:
Anyone that has write access on a ClusterExtension can use any existing service account in the cluster to manage bundle content. Currently, the only bundle content that is possible to install is content that is available in a ClusterCatalog.

So in order to perform privilege escalation, someone would need:
1. Write access to the ClusterExtension API
2. Write access to the ClusterCatalog API OR an existing ClusterCatalog that contains escalated privileges that the actor could make use of.

In general, these APIs are considered privilieged APIs because they are meant to manage cluster extensions, which often contain CRDs and RBAC.

This is an improvement over OLMv0, which itself had cluster-admin permissions.

Perhaps there is a way to introduce authorization during ClusterExtension admission that compares user credentials with service account credentials to ensure that the user has a superset of the permissions of the service account? However, I'm not sure if it would be possible to make this work with arbitrary authorizers configured on the apiserver.

Reply:
Some admission time privilege escalation checks, using subject access review would be awesome, I'm not sure how complex that is to integrate though.

So for now, the basis of the security here is that we document that these are privileged, and if you happen to delegate write permissions to ClusterExtensions, you better be sure you know who you're giving that permission to? ie, reserve to cluster admins?

Reply:
>Some admission time privilege escalation checks, using subject access review would be awesome, I'm not sure how complex that is to integrate though.

We are planning to work on a pre-flight check that evaluates the permissions of the provided ServiceAccount to see if it is sufficient for installation before attempting to do the installation. This will also help show the privilege escalation scenario by failing the pre-flight check if the installation requires more permissions than the previous one.

I had initially contributed some checks of this nature to kapp here: https://github.com/carvel-dev/kapp/tree/develop/pkg/kapp/permissions

I think we are looking into whether or not we can re-use that logic for our needs in OLM.

>So for now, the basis of the security here is that we document that these are privileged, and if you happen to delegate write permissions to ClusterExtensions, you better be sure you know who you're giving that permission to? ie, reserve to cluster admins?

I believe so.